cast/src/tx.rs

```diff
@@ -0,0 +1,360 @@
+use ethers_core::{
+    abi::Function,
+    types::{
+        transaction::eip2718::TypedTransaction, Chain, Eip1559TransactionRequest, NameOrAddress,
+        TransactionRequest, H160, U256,
+    },
+};
+use ethers_providers::Middleware;
+
+use eyre::{eyre, Result};
+use foundry_utils::{encode_args, get_func, get_func_etherscan};
+use futures::future::join_all;
+
+use crate::strip_0x;
+
+pub struct TxBuilder {
+    to: H160,
+    chain: Chain,
+    tx: TypedTransaction,
+    func: Option<Function>,
+    etherscan_api_key: Option<String>,
+}
+
+///
+/// Transaction builder
+/// ```
+/// async fn foo() -> eyre::Result<()> {
+///   use ethers_core::types::{Chain, U256};
+///   use cast::{TxBuilder};
+///   let provider = ethers_providers::test_provider::MAINNET.provider();
+///   let mut builder = TxBuilder::new(&provider, "a.eth", "b.eth", Chain::Mainnet, false).await?;
+///   builder
+///       .gas(Some(U256::from(1)));
+///   let (tx, _) = builder.build();
+///   Ok(())
+/// }
+/// ```
+impl TxBuilder {
+    ///
+    /// Create a new TxBuilder
+    /// `provider` - provider to use
+    /// `from` - 'from' field. Could be an ENS name
+    /// `to` - `to`. Could be a ENS
+    /// `chain` - chain to construct the tx for
+    /// `legacy` - use type 1 transaction
+    pub async fn new<M: Middleware, F: Into<NameOrAddress>, T: Into<NameOrAddress>>(
+        provider: &M,
+        from: F,
+        to: T,
+        chain: Chain,
+        legacy: bool,
+    ) -> Result<TxBuilder> {
+        let from_addr = resolve_ens(provider, from).await?;
+        let to_addr = resolve_ens(provider, foundry_utils::resolve_addr(to, chain)?).await?;
+
+        let tx: TypedTransaction = if chain.is_legacy() || legacy {
+            TransactionRequest::new().from(from_addr).to(to_addr).into()
+        } else {
+            Eip1559TransactionRequest::new().from(from_addr).to(to_addr).into()
+        };
+
+        Ok(TxBuilder { to: to_addr, chain, tx, func: None, etherscan_api_key: None })
+    }
+    ///
+    /// Set gas for tx
+    pub fn set_gas(&mut self, v: U256) -> &mut TxBuilder {
+        self.tx.set_gas(v);
+        self
+    }
+
+    ///
+    /// Set gas for tx, if `v` is not None
+    pub fn gas(&mut self, v: Option<U256>) -> &mut TxBuilder {
+        match v {
+            Some(x) => self.set_gas(x),
+            None => self,
+        }
+    }
+
+    ///
+    /// Set gas price
+    pub fn set_gas_price(&mut self, v: U256) -> &mut TxBuilder {
+        self.tx.set_gas_price(v);
+        self
```

Comment:
nit: changing this and `gas_price` makes it cleaner imo (feel free to ignore)

```suggestion
        self.gas_price(Some(v))
```